jupyter-extension/jupyterlab_pachyderm/pfs_manager.py

```diff
@@ -214,16 +219,325 @@ def get(self, path, content=True, type=None, format=None) -> dict:
+    that file would exist serialized on disk at:
+    /pfs_datum/default_test_master/foo/bar/file.txt
+  - Both files and directories that are mounted are stored this way. What should
+    be displayed, however, varies depending on if a file or directory is mounted.
+  - For files, both the file itself as well as any parent directories of the file
+    should exist in /pfs_datum
+  - For directories, any files or directories within that directory should be
+    mounted as well.
+"""
+class DatumManager(FileContentsManager):
+    # currently unsupported stuff (unclear if needed or not):
+    #  - crossing repo with itself
+    #  - renaming repo level directories
+
+    def __init__(self, client: Client, **kwargs):
+        self._client = client
+        self._datum_list = list()
+        self._FILEINFO_DIR = "/pfs_datum"
+        self._dirs = set()
+        self._datum_index = 0
+        self._mount_time = datetime.datetime.min
+        os.makedirs(self._FILEINFO_DIR, exist_ok=True)
+        # TODO: expose the apis for mounting/cycling datums
+        # input = {'input': {'pfs': {'repo': 'test', 'glob': '/file*'}}}
+        # self.mount_datums(input_dict=input)
+        # self.next_datum()
+
+    # TODO: don't ignore name in the input spec
+    # right now, when a repo is mounted as part of mounting datum(s), the
+    # toplevel directory for the repo is project_repo_branch. we should
+    # update this to respect the name in the pfsinput objects being supplied
+    # to the mount_datums api. however, this is not that simple as the name
+    # can't simply be looked up in the datuminfo. we will need to have some
+    # mapping to track which files within a datum belong to which name.
+    def mount_datums(self, input_dict: dict):
+        input = pps.Input().from_dict(input_dict["input"])
+        datum = self._client.pps.list_datum(input=input)
+        self._datum_list = [d for d in datum]
+        self._datum_index = 0
+        self._mount_time = datetime.datetime.now()
+        if len(self._datum_list) == 0:
+            raise ValueError("input produced no datums to mount")
+        self._update_mount()
+
+    def next_datum(self):
+        self._datum_index = (self._datum_index + 1) % len(self._datum_list)
+        self._update_mount()
+
+    def prev_datum(self):
+        if self._datum_index == 0:
+            self._datum_index = len(self._datum_list)
+        self._datum_index = self._datum_index - 1
+        self._update_mount()
+
+    def _get_fileinfo_path(self, fileinfo: pfs.FileInfo) -> Path:
+        project = fileinfo.file.commit.repo.project.name
+        branch = fileinfo.file.commit.branch.name
+        repo = fileinfo.file.commit.repo.name
+        toplevel = f"{project}_{repo}_{branch}"
+        return Path(self._FILEINFO_DIR, toplevel, fileinfo.file.path.strip("/"))
+
+    def _update_mount(self):
+        os.system(f"rm -rf {self._FILEINFO_DIR}/*")
```

Comment:
sorry, i should have documented this better - `_dirs` is a set that keeps track of directories mounted in the datum. there is a distinction between directories being mounted and directories that exist in our file structure because one of their children were mounted in that the former requires all of their children to also be displayed in the contents manager while the latter does not. `_dirs` specifically keeps track of the former so that we can more quickly verify whether a file that may be a child of one of these exists. that being said, i can replace the `os.system` call with `shutil.rmtree` if that is the preferred way of removing directories on disk.